[Inductor] Fix max_autotune BMM correctness with dynamic OpenMP threads

dumko2001 opened the pull request:
# Summary
This PR fixes a silent correctness bug in `max_autotune` BMM kernels when the OpenMP thread count is dynamic (e.g., reduced by external libraries like `cv2`).

Previously, the generated C++ code assumed a 1-to-1 mapping between physical threads and work items (`tid = omp_get_thread_num()`). If OpenMP provided fewer threads than requested, tasks for higher `tid`s were simply skipped, leading to wrong results.

This change switches from a raw `#pragma omp parallel` block to `#pragma omp parallel for`. This ensures OpenMP automatically distributes the loop iterations (work items) across available threads, guaranteeing all work is computed regardless of the actual thread count.

# Test Plan
Verified that the generated C++ template in `torch/_inductor/codegen/cpp_gemm_template.py` now uses:
```cpp
#pragma omp parallel for num_threads(...)
for (int64_t tid = 0; tid < ...; tid++) { ... }
This preserves existing memory barrier/cache behavior while fixing the task distribution logic.
```
Fixes
Fixes #168965

# Performance Note
This change switches the scheduling model to `omp parallel for`. While this fixes the correctness bug, I could not verify if this introduces any scheduling overhead on standard benchmarks due to hardware limitations. 
It would be helpful if a reviewer could run a standard Inductor benchmark to ensure no performance regression.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

desertfire reviewed (commented):
Thanks for the fix! Can you add a unit test for this?

dumko2001 commented:
@pytorchbot label "topic: not user facing"

desertfire reviewed (commented):
To fix lint error, please refer to https://github.com/pytorch/pytorch/wiki/lintrunner

dumko2001 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2,12 +2,14 @@
 import platform
 import sys
-import unittest
+import unittes
```
yep , I did the lint runner thing

dumko2001 commented:
@desertfire,

I've isolated the core regression to the OpenMP parallelization structure in cpp_gemm_template.py.

The issue is that the logic inside the loop requires synchronization (likely a reduction/barrier for K-slicing) which is semantically invalid when wrapped in a single #pragma omp parallel for. This is causing the multiple CI failures (deadlocks/correctness errors).

The fix requires a structural restructuring of the parallel region. We need to switch to a model that allows for explicit synchronization, such as:

Option: Compute/Reduce Split (Two sequential #pragma omp for blocks inside one #pragma omp parallel block).
Option: Single parallel for Loop (if partitioning can be made purely thread-independent).
Can you advise which parallel structure (Split-Loop vs. Thread-Centric Parallel Region) is preferred for the Inductor codebase to handle this K-slicing reduction?

dumko2001 commented:
@desertfire,

Could you please manually remove the topic: not user facing label?

The PR is a correctness fix that should be included in the release notes, Thank you!

jgong5 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -5834,6 +5836,60 @@ def fn(x):
+                    torch.testing.assert_close(actual, expected)
+            finally:
+                # Reset dynamic threading
+                omp.omp_set_dynamic(0)
```
Should we restore the value to the original `omp_get_dynamic`?

dumko2001 replied:
Fixed. Now saves `original_dynamic = omp.omp_get_dynamic()` before modification and restores it in the `finally` block.

jgong5 commented on torch/_inductor/codegen/cpp_grouped_gemm_template.py:
```diff
@@ -48,7 +48,9 @@
     #pragma omp parallel num_threads({{num_threads}})
     {
-        {{ template.codegen_multi_threads_params()|indent(8, false) }}
+        #pragma omp for
```
ditto: `#pragma omp for schedule(static, 1)`

jgong5 reviewed (changes requested):
As was mentioned in the PR description, we'd better check the perf impact on the changes. @CaoE can we do that?

dumko2001 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -5834,6 +5836,60 @@ def fn(x):
+            self.skipTest("OpenMP library not found")
+
+        try:
+            omp = ctypes.CDLL(lib_path)
```
Fixed. Now using `ctypes.CDLL("", ctypes.RTLD_GLOBAL)` to access the already-loaded OpenMP library.

jgong5 reviewed (changes requested):
Some nits left, please check. Also, please avoid force-push since it makes reviewing incremental changes harder.

jgong5 commented:
@dumko2001 Could you please resolve conflicts first?

jgong5 reviewed (changes requested):
Please resolve the conflict and tag me for review again.

dumko2001 committed:
[Inductor] Fix max_autotune BMM correctness with dynamic OpenMP threads

When omp_set_dynamic(1) is active (e.g. after importing cv2), OpenMP may
spawn fewer threads than requested by num_threads(N).  The old code used
omp_get_thread_num() to index into per-thread work, so any thread ID that
was never created left its slice of the output uncomputed, producing wrong
results.

Replace the omp_get_thread_num() pattern with #pragma omp for
schedule(static, 1) inside an explicit for(tid=0; tid<N; tid++) loop.
OpenMP's work-sharing guarantees that every iteration runs regardless of
how many actual threads are spawned, eliminating the correctness bug.

Also remove the unused DTYPE_TO_CPP import from cpp_grouped_gemm_template
and add a regression test that exercises the bug by calling
omp_set_dynamic(1) before running a compiled bmm.

desertfire commented:
@dumko2001 , will you continue to finish landing this PR?

dumko2001 commented:
Yea sorry will get on it today

dumko2001 committed:
Fix nested OpenMP worksharing in GEMM template

dumko2001 commented:
@desertfire done , can u run the ci

dumko2001 commented:
@desertfire I have rebased since lint was failing. can u run ci again

dumko2001 committed:
[Inductor] Fix micro-gemm state scope in GEMM templates

jgong5 commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge -i
